lib/probe.ml

```diff
@@ -0,0 +1,352 @@
+let is_confirmed state name =
+  UniqueSet.mem name state.names_confirmed
+
+let stop state =
+  { state with stage = ProbeStopped }
+
+let is_first_complete state = state.first_done
+
+let prepare_probe names db =
+  (* Build a list of questions *)
+  let questions = List.map (fun name -> Packet.({
+      q_name = name;
+      q_type = Q_ANY_TYP;
+      q_class = Q_IN;
+      q_unicast = Q_mDNS_Unicast;  (* request unicast response as per RFC 6762 section 8.1 para 6 *)
+    })) names
+  in
+  (* Reuse Query.answer_multiple to get the records that we need for the authority section *)
+  let answer = Query.answer_multiple ~dnssec:false ~mdns:true questions db.Loader.trie in
+  let rrs = List.filter (fun answer -> List.mem answer.Packet.name names) answer.Query.answer in
+  if rrs = [] then
+    (* There are no unique records to probe for *)
+    None
+  else
+    (* I don't know whether the cache flush bit needs to be set in the authority RRs, but seems logical *)
+    let authorities = List.map (fun rr -> { rr with Packet.flush = true }) rrs in
+    let detail = Packet.({ qr=Query; opcode=Standard; aa=false; tc=false; rd=false; ra=false; rcode=NoError; }) in
+    let query = Packet.({ id=0; detail; questions; answers=[]; authorities; additionals=[]; }) in
+    Some (query, rrs)
+
+(* Initiates the first probe *)
+let do_probe state =
+  let begin_probe () =
+    match prepare_probe (UniqueSet.elements state.names_pending) state.db with
+    | None ->
+      (* Nothing to do right now *)
+      ({ state with stage = ProbeIdle }, Nothing)
+    | Some (packet, rrs) ->
+      (* Send the probe *)
+      (* TODO: probes should be per-link if there are multiple NICs *)
+      let datagram = (packet,multicast_ip,5353) in
+      ({
+        state with
+        stage = SendingProbe { datagram; num=FirstProbe; rrs };
+        names_pending = UniqueSet.empty;
+        names_probing = state.names_pending;
+      }, ToSend datagram)
+  in
+  match state.stage with
+  | ProbeIdle
+    ->
+    begin_probe ()
+  | NeedRestart (DoProbe, delay)
+    ->
+    if delay = 0.0 then
+      begin_probe ()
+    else
+      ({ state with stage = DelayBeforeRestart }, Delay delay)
+  | SendingProbe _
+  | DelayAfterSendingProbe _
+  | NeedRestart (AfterSend, _)
+  | NeedRestart (AfterDelay, _)
+  | DelayBeforeRestart
+    ->
+    (state, NotReady)
+  | ProbeStopped
+    ->
+    (state, Stop)
+
+let restart_later state delay =
+  match state.stage with
+  | SendingProbe _ ->
+    { state with stage = NeedRestart (AfterSend, delay) }
+  | DelayAfterSendingProbe _ ->
+    (* Delays a bit longer than needed *)
+    { state with stage = NeedRestart (AfterDelay, delay) }
+  | ProbeIdle ->
+    { state with stage = NeedRestart (DoProbe, delay) }
+  | NeedRestart _
+  | DelayBeforeRestart
+  | ProbeStopped ->
+    (* No change *)
+    state
+
+let on_send_complete state =
+  match state.stage with
+  | SendingProbe probing
+    ->
+    (* Fixed delay of 250 ms *)
```

Comment:
Do these timeouts ever need to be changed (e.g. in a test harness)? Or is the intention that the callback that receives `Delay` can change them?  I'm just wondering if the 250ms value should be an optional argument to `on_send_complete` rather than hardcoded here.